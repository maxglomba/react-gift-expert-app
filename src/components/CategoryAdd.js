import React, { useState } from 'react';
import PropTypes from 'prop-types';

const CategoryAdd = ({ categories, setCategories }) => {

    const [inputValue, setInputValue] = useState('');

    const handleChange = (e) => {
        setInputValue(e.target.value);
    }

    const handleSubmit = (e) => {
        e.preventDefault();
        if (!categories.includes(inputValue.toLowerCase()) && inputValue.trim().length > 2) {
            setCategories( cats => [inputValue, ...categories]);
            setInputValue('');
        }
    }
    return (
        <form onSubmit={handleSubmit}>
            <p hidden>{inputValue}</p>
            <input
                type='text'
                value={inputValue}
                placeholder='Ingrese su busqueda'
                id='searchText'
                onChange={handleChange}
            />
        </form>
    )


}

CategoryAdd.propTypes = {
    categories: PropTypes.array.isRequired,
    setCategories: PropTypes.func.isRequired,
}

export default CategoryAdd;