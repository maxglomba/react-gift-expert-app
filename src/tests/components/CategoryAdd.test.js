import React from "react";
import "@testing-library/jest-dom";
import { shallow } from "enzyme"
import CategoryAdd from "../../components/CategoryAdd"

describe('Pruebas en componente <CategoryAdd />', () => { 
    const categories = [];
    const setCategories = jest.fn();
    let wrapper =  shallow(<CategoryAdd categories={categories} setCategories={setCategories} />);
    beforeEach( () => {
        jest.clearAllMocks();
        wrapper = shallow(<CategoryAdd categories={categories} setCategories={setCategories} />);
    })

    test('debe renderizarse correctamente', () => { 
        expect(wrapper).toMatchSnapshot();
    })

    test('debe cambiar el state en base a la caja de texto', () => { 
        const input = wrapper.find('input');
        const value = 'Hola mundo';
        input.simulate('change', { target:{value} });
        expect(wrapper.find('p').text().trim()).toBe(value);
    })

    test('NO debe de postear la informacion en submit si tiene menos de 2 caracteres', () => { 
        wrapper.find('form').simulate('submit', { preventDefault(){}});
        expect( setCategories ).not.toHaveBeenCalled()

     })

     test('debe de llamar el setCategories y limpiar la caja de texto', () => { 
         //simulo llenar el input
        const input = wrapper.find('input');
        const value = 'Hola mundo';
        input.simulate('change', { target:{value} });
        expect(wrapper.find('p').text().trim()).toBe(value);

        //simulo el submit
        wrapper.find('form').simulate('submit', { preventDefault(){}});
        expect( setCategories ).toHaveBeenCalledTimes(1);
        expect( setCategories ).toHaveBeenCalledWith( expect.any(Function));

        //controlo que se haya vaciado el input
        expect(wrapper.find('input').prop('value')).toBe('');
     })

})
