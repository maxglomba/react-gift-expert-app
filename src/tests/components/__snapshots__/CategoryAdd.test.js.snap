// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en componente <CategoryAdd /> debe renderizarse correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <p
    hidden={true}
  />
  <input
    id="searchText"
    onChange={[Function]}
    placeholder="Ingrese su busqueda"
    type="text"
    value=""
  />
</form>
`;
